cirq-core/cirq/ops/gateset.py

```diff
@@ -0,0 +1,352 @@
+        return self._gate
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def description(self) -> str:
+        return self._description
+
+    def predicate(self, gate: raw_types.Gate) -> bool:
+        """Checks whether `cirq.Gate` instance `g` belongs to this GateFamily.
+
+        The default predicate depends on the gate family initialization type:
+            a) Instance Family: `gate == self.gate`.
+            b) Type Family: `isinstance(gate, self.gate)`.
+
+        Args:
+            gate: `cirq.Gate` instance which should be checked for containment.
+        """
+        return (
+            gate == self.gate
+            if isinstance(self.gate, raw_types.Gate)
+            else isinstance(gate, self.gate)
+        )
+
+    def __contains__(self, item: Union[raw_types.Gate, raw_types.Operation]) -> bool:
+        if isinstance(item, raw_types.Operation):
+            if item.gate is None:
+                return False
+            item = item.gate
+        return self.predicate(item)
+
+    def __str__(self) -> str:
+        return f'{self.name}\n{self.description}'
+
+    def __repr__(self) -> str:
+        return (
+            f'cirq.GateFamily(gate={self._gate_str(repr)},'
+            f'name="{self.name}", '
+            f'description="{self.description}")'
+        )
+
+    def _value_equality_values_(self) -> Any:
+        # `isinstance` is used to ensure the a gate type and gate instance is not compared.
+        return isinstance(self.gate, raw_types.Gate), self.gate, self.name, self.description
+
+
+@value.value_equality()
+class Gateset:
+    """Gatesets represent a collection of `cirq.GateFamily` objects.
+
+    Gatesets are useful for
+        a) Describing the set of allowed gates in a human readable format
+        b) Validating a given gate / optree against the set of allowed gates
+
+    Gatesets rely on the underlying `cirq.GateFamily` for both description and
+    validation purposes.
+    """
+
+    def __init__(
+        self,
+        *gates: Union[Type[raw_types.Gate], raw_types.Gate, GateFamily],
+        name: Optional[str] = None,
+        unroll_circuit_op: bool = True,
+        accept_global_phase: bool = True,
+    ) -> None:
+        """Init Gateset.
+
+        Accepts a list of gates, each of which should be either
+            a) `cirq.Gate` subclass
+            b) `cirq.Gate` instance
+            c) `cirq.GateFamily` instance
+
+        `cirq.Gate` subclasses and instances are converted to the default
+        `cirq.GateFamily(gate=g)` instance and thus a default name and
+        description is populated.
+
+        Args:
+            *gates: A list of `cirq.Gate` subclasses / `cirq.Gate` instances /
+            `cirq.GateFamily` instances to initialize the Gateset.
+            name: (Optional) Name for the Gateset. Useful for description.
+            unroll_circuit_op: If True, `cirq.CircuitOperation` is recursively
```

Comment:
Looking through this we aren't really unrolling vs not unrolling circuit operation anymore. We are more like "supporting" vs "not supporting" CircuitOperation. Can we change this to `include/accept_circuit_op` or something similar ?

Reply:
As discussed offline, `include` and `accept` can be more confusing and `unroll` seems to be to indicative of what's happening. Resolving this now.